cocos/ui/UIRichText.cpp

```diff
@@ -1609,110 +1646,94 @@ void RichText::handleTextRenderer(const std::string& text, const std::string& fo
                                   const Color3B& glowColor)
 {
-    auto fileExist = FileUtils::getInstance()->isFileExist(fontName);
-    Label* textRenderer = nullptr;
-    if (fileExist)
-    {
-        textRenderer = Label::createWithTTF(text, fontName, fontSize);
-    } 
-    else
-    {
-        textRenderer = Label::createWithSystemFont(text, fontName, fontSize);
-    }
-    if (flags & RichElementText::ITALICS_FLAG)
-        textRenderer->enableItalics();
-    if (flags & RichElementText::BOLD_FLAG)
-        textRenderer->enableBold();
-    if (flags & RichElementText::UNDERLINE_FLAG)
-        textRenderer->enableUnderline();
-    if (flags & RichElementText::STRIKETHROUGH_FLAG)
-        textRenderer->enableStrikethrough();
-    if (flags & RichElementText::URL_FLAG)
-        textRenderer->addComponent(ListenerComponent::create(textRenderer,
-                                                             url,
-                                                             std::bind(&RichText::openUrl, this, std::placeholders::_1)));
-    if (flags & RichElementText::OUTLINE_FLAG) {
-        textRenderer->enableOutline(Color4B(outlineColor), outlineSize);
-    }
-    if (flags & RichElementText::SHADOW_FLAG) {
-        textRenderer->enableShadow(Color4B(shadowColor), shadowOffset, shadowBlurRadius);
-    }
-    if (flags & RichElementText::GLOW_FLAG) {
-        textRenderer->enableGlow(Color4B(glowColor));
-    }
-
-    float textRendererWidth = textRenderer->getContentSize().width;
-    _leftSpaceWidth -= textRendererWidth;
-    if (_leftSpaceWidth < 0.0f)
+    bool fileExist = FileUtils::getInstance()->isFileExist(fontName);
+    RichText::WrapMode wrapMode = static_cast<RichText::WrapMode>(_defaults.at(KEY_WRAP_MODE).asInt());
+
+    // split text by \n
+    std::stringstream ss(text);
+    std::string currentText;
+    size_t realLines = 0;
+    while (std::getline(ss, currentText, '\n'))
     {
-        int leftLength = 0;
-        if (static_cast<RichText::WrapMode>(_defaults.at(KEY_WRAP_MODE).asInt()) == WRAP_PER_WORD)
-            leftLength = findSplitPositionForWord(textRenderer, text);
-        else
-            leftLength = findSplitPositionForChar(textRenderer, text);
-
-        //The minimum cut length is 1, otherwise will cause the infinite loop.
-//        if (0 == leftLength) leftLength = 1;
-        std::string leftWords = Helper::getSubStringOfUTF8String(text, 0, leftLength);
-        int rightStart = leftLength;
-        if (std::isspace(text[rightStart], std::locale()))
-            rightStart++;
-        std::string cutWords = Helper::getSubStringOfUTF8String(text, rightStart, text.length() - leftLength);
-        if (leftLength > 0)
+        if (realLines > 0)
         {
-            Label* leftRenderer = nullptr;
-            if (fileExist)
+            addNewLine();
+            _defaultHeights.back() = fontSize;
```

Comment:
why heights vector record font size?

Reply:
This is for continuous new line.
eg："abc\n\n123" should display like this:
abc

123

Reply:
may be should use another name for the member variable, the name is `_defaultHeights `, but the element value is font size, it is hard to understand.

Reply:
How about `_emptyLineHeights`

Reply:
I don't quite understand why height is equal to font size. When i reviewed the codes, it seems font size is used as height.

Reply:
Suppose a RichElementText's text is "abc\n\n123" with fontSize 10, and the RichText’s fontSize is 20.
Then it will display like this:
abc

123

Now,  the gap between two lines should 10, right?